ingester/src/data.rs

```diff
@@ -868,6 +931,25 @@ impl DataBuffer {
         QueryableBatch::new(table_name, data, tombstones)
     }
 
+    /// Returns all exisitng snapshots plus data in the buffer
+    /// This only read data. Data in the buffer will be kept in the buffer
+    pub fn buffer_and_snapshots(&self) -> Result<Vec<Arc<SnapshotBatch>>> {
+        let mut snapshots = vec![];
+        // copy the buffer to a snapshot
+        let buffer_sanpshot = self
+            .copy_buffer_to_snapshot()
+            .context(BufferToSnapshotSnafu)?;
+        if let Some(snapshot) = buffer_sanpshot {
+            snapshots.push(snapshot);
+        }
+
+        // other existing snapshots
+        let mut other_snapshots: Vec<_> = self.snapshots.iter().map(Arc::clone).collect();
+        snapshots.append(&mut other_snapshots);
```

Comment:
This can use `extend` and avoid the allocation of the `Vec`:

```suggestion
        snapshots.extend(self.snapshots.iter().cloned());
```